File: packages/monaco/src/providers/completion.ts
// Monaco `CompletionItemProvider` backed by `Project::completion`.
// Maps the IDE-shape `CompletionList` from `@greycat/analyzer` to
// Monaco's `CompletionList`. The IDE `CompletionItemKind` enum is
// a strict subset of Monaco's, so the per-kind mapping is total.

import type * as MonacoNs from "monaco-editor";
import type { Project, CompletionItem as IdeCompletionItem } from "@greycat/analyzer";
import { CompletionItemKind } from "@greycat/analyzer";

export function registerCompletion(
  monaco: typeof MonacoNs,
  project: Project,
  languageId: string,
): MonacoNs.IDisposable {
  return monaco.languages.registerCompletionItemProvider(languageId, {
    triggerCharacters: [".", ":", "@", '"'],
    provideCompletionItems(model, position) {
      const uri = model.uri.toString();
      const list = project.completion(uri, position.lineNumber - 1, position.column - 1);
      if (!list) {
        return { suggestions: [] };
      }
      return {
        incomplete: list.is_incomplete,
        suggestions: list.items.map((item) => toMonaco(monaco, item, position)),
      };
    },
  });
}

function toMonaco(
  monaco: typeof MonacoNs,
  item: IdeCompletionItem,
  position: MonacoNs.Position,
): MonacoNs.languages.CompletionItem {
  const range: MonacoNs.IRange = item.text_edit
    ? {
        startLineNumber: item.text_edit.range.start.line + 1,
        startColumn: item.text_edit.range.start.character + 1,
        endLineNumber: item.text_edit.range.end.line + 1,
        endColumn: item.text_edit.range.end.character + 1,
      }
    : {
        startLineNumber: position.lineNumber,
        startColumn: position.column,
        endLineNumber: position.lineNumber,
        endColumn: position.column,
      };
  return {
    label: item.label,
    kind: kindToMonaco(monaco, item.kind),
    insertText: item.text_edit?.new_text ?? item.insert_text ?? item.label,
    insertTextRules:
      item.insert_text_format === 1
        ? monaco.languages.CompletionItemInsertTextRule.InsertAsSnippet
        : undefined,
    range,
    sortText: item.sort_text,
    filterText: item.filter_text,
    detail: item.detail,
    documentation: item.documentation ? { value: item.documentation, isTrusted: true } : undefined,
    additionalTextEdits: item.additional_text_edits?.map((edit) => ({
      range: {
        startLineNumber: edit.range.start.line + 1,
        startColumn: edit.range.start.character + 1,
        endLineNumber: edit.range.end.line + 1,
        endColumn: edit.range.end.character + 1,
      },
      text: edit.new_text,
    })),
  };
}

function kindToMonaco(
  monaco: typeof MonacoNs,
  kind: CompletionItemKind | undefined,
): MonacoNs.languages.CompletionItemKind {
  if (kind === undefined) {
    return monaco.languages.CompletionItemKind.Text;
  }
  switch (kind) {
    case CompletionItemKind.Function:
      return monaco.languages.CompletionItemKind.Function;
    case CompletionItemKind.Method:
      return monaco.languages.CompletionItemKind.Method;
    case CompletionItemKind.Variable:
      return monaco.languages.CompletionItemKind.Variable;
    case CompletionItemKind.Field:
      return monaco.languages.CompletionItemKind.Field;
    case CompletionItemKind.Class:
      return monaco.languages.CompletionItemKind.Class;
    case CompletionItemKind.Enum:
      return monaco.languages.CompletionItemKind.Enum;
    case CompletionItemKind.EnumMember:
      return monaco.languages.CompletionItemKind.EnumMember;
    case CompletionItemKind.Constant:
      return monaco.languages.CompletionItemKind.Constant;
    case CompletionItemKind.Module:
      return monaco.languages.CompletionItemKind.Module;
    case CompletionItemKind.Folder:
      return monaco.languages.CompletionItemKind.Folder;
    case CompletionItemKind.Keyword:
      return monaco.languages.CompletionItemKind.Keyword;
    case CompletionItemKind.Text:
      return monaco.languages.CompletionItemKind.Text;
    case CompletionItemKind.TypeParameter:
      return monaco.languages.CompletionItemKind.TypeParameter;
    default: {
      // Exhaustive — if a new variant lands in
      // `@greycat/analyzer`'s `CompletionItemKind` and isn't wired up
      // here, this branch fails the compile via the `never` check.
      const _exhaustive: never = kind;
      return _exhaustive;
    }
  }
}

File: greycat-analyzer-server/src/capabilities/completion.rs
//! Thin converter from the IDE-shape `analysis::ide::completion::*`
//! ADTs to `lsp_types::CompletionList` + `CompletionItem`. The dispatcher,
//! scope walks, member discovery, lib-version placeholder + resolution
//! all live in the analysis crate so the CLI and wasm bridge share the
//! same source.

use greycat_analyzer_analysis::ide::completion::{
    CompletionItem as IdeCompletionItem, CompletionItemKind as IdeCompletionItemKind,
    CompletionItemLabelDetails as IdeCompletionItemLabelDetails,
    CompletionList as IdeCompletionList, InsertTextFormat as IdeInsertTextFormat,
    LibVersionPayload, completion_with_project as completion_inner,
    extract_lib_version_placeholder as extract_lib_version_placeholder_inner,
    resolve_library_version_completion as resolve_library_version_completion_inner,
};
use greycat_analyzer_analysis::ide::types::{
    Position as IdePosition, Range as IdeRange, TextEdit as IdeTextEdit,
};
use greycat_analyzer_analysis::project::ProjectAnalysis;
use greycat_analyzer_core::{SourceEncoding, registry::RegistryFetcher};
use greycat_analyzer_syntax::tree_sitter;
use lsp_types::{
    CompletionItem, CompletionItemKind, CompletionItemLabelDetails, CompletionList,
    CompletionTextEdit, Documentation, InsertTextFormat, MarkupContent, MarkupKind, Position,
    Range, TextEdit, Uri,
};

#[allow(clippy::too_many_arguments)]
pub fn completion_with_project(
    text: &str,
    root: tree_sitter::Node<'_>,
    pos: Position,
    uri: &Uri,
    project: &ProjectAnalysis,
    project_root: Option<&std::path::Path>,
    encoding: SourceEncoding,
) -> Option<CompletionList> {
    completion_inner(text, root, pos, uri, project, project_root, encoding).map(list_to_lsp)
}

/// LSP-side projection of the IDE placeholder extractor. Reads the
/// LSP-shape `CompletionList` by walking back through its single item's
/// `data` field — same JSON shape on both sides.
pub fn extract_lib_version_placeholder(list: &CompletionList) -> Option<LibVersionPayload> {
    // Round-trip through the IDE shape: the JSON `data` payload is the
    // same shape regardless of the carrier struct, so we reuse the
    // analysis-side extractor.
    let ide_list = lsp_list_to_ide(list);
    extract_lib_version_placeholder_inner(&ide_list)
}

pub fn resolve_library_version_completion(
    payload: &LibVersionPayload,
    fetcher: &dyn RegistryFetcher,
) -> CompletionList {
    list_to_lsp(resolve_library_version_completion_inner(payload, fetcher))
}

fn list_to_lsp(list: IdeCompletionList) -> CompletionList {
    CompletionList {
        is_incomplete: list.is_incomplete,
        items: list.items.into_iter().map(item_to_lsp).collect(),
    }
}

fn lsp_list_to_ide(list: &CompletionList) -> IdeCompletionList {
    IdeCompletionList {
        is_incomplete: list.is_incomplete,
        items: list.items.iter().map(lsp_item_to_ide).collect(),
    }
}

fn item_to_lsp(item: IdeCompletionItem) -> CompletionItem {
    CompletionItem {
        label: item.label,
        kind: item.kind.map(kind_to_lsp),
        detail: item.detail,
        documentation: item.documentation.map(|markdown| {
            Documentation::MarkupContent(MarkupContent {
                kind: MarkupKind::Markdown,
                value: markdown,
            })
        }),
        insert_text: item.insert_text,
        insert_text_format: item.insert_text_format.map(insert_text_format_to_lsp),
        sort_text: item.sort_text,
        filter_text: item.filter_text,
        label_details: item.label_details.map(label_details_to_lsp),
        text_edit: item
            .text_edit
            .map(|te| CompletionTextEdit::Edit(text_edit_to_lsp(te))),
        additional_text_edits: item
            .additional_text_edits
            .map(|edits| edits.into_iter().map(text_edit_to_lsp).collect()),
        data: item.data,
        ..Default::default()
    }
}

fn lsp_item_to_ide(item: &CompletionItem) -> IdeCompletionItem {
    IdeCompletionItem {
        label: item.label.clone(),
        kind: None,
        detail: item.detail.clone(),
        documentation: None,
        insert_text: item.insert_text.clone(),
        insert_text_format: None,
        sort_text: item.sort_text.clone(),
        filter_text: item.filter_text.clone(),
        label_details: None,
        text_edit: None,
        additional_text_edits: None,
        data: item.data.clone(),
    }
}

fn kind_to_lsp(kind: IdeCompletionItemKind) -> CompletionItemKind {
    match kind {
        IdeCompletionItemKind::Function => CompletionItemKind::FUNCTION,
        IdeCompletionItemKind::Method => CompletionItemKind::METHOD,
        IdeCompletionItemKind::Variable => CompletionItemKind::VARIABLE,
        IdeCompletionItemKind::Field => CompletionItemKind::FIELD,
        IdeCompletionItemKind::Class => CompletionItemKind::CLASS,
        IdeCompletionItemKind::Enum => CompletionItemKind::ENUM,
        IdeCompletionItemKind::EnumMember => CompletionItemKind::ENUM_MEMBER,
        IdeCompletionItemKind::Constant => CompletionItemKind::CONSTANT,
        IdeCompletionItemKind::Module => CompletionItemKind::MODULE,
        IdeCompletionItemKind::Folder => CompletionItemKind::FOLDER,
        IdeCompletionItemKind::Keyword => CompletionItemKind::KEYWORD,
        IdeCompletionItemKind::Text => CompletionItemKind::TEXT,
        IdeCompletionItemKind::TypeParameter => CompletionItemKind::TYPE_PARAMETER,
    }
}

fn insert_text_format_to_lsp(fmt: IdeInsertTextFormat) -> InsertTextFormat {
    match fmt {
        IdeInsertTextFormat::PlainText => InsertTextFormat::PLAIN_TEXT,
        IdeInsertTextFormat::Snippet => InsertTextFormat::SNIPPET,
    }
}

fn label_details_to_lsp(d: IdeCompletionItemLabelDetails) -> CompletionItemLabelDetails {
    CompletionItemLabelDetails {
        detail: d.detail,
        description: d.description,
    }
}

fn text_edit_to_lsp(te: IdeTextEdit) -> TextEdit {
    TextEdit {
        range: range_to_lsp(te.range),
        new_text: te.new_text,
    }
}

fn range_to_lsp(r: IdeRange) -> Range {
    Range {
        start: pos_to_lsp(r.start),
        end: pos_to_lsp(r.end),
    }
}

fn pos_to_lsp(p: IdePosition) -> Position {
    Position {
        line: p.line,
        character: p.character,
    }
}
